### DISCUSSION TITLE
Deploy to remote server using SSH fails on "Host key verification failed"

### DISCUSSION THREAD

**Discussion by felixmosh** [community/community#26564] [Actions]
<p>I’ve Generated new SSH keys (private &amp; public), stored the private SSH key as a secret and added the public one to <code>.ssh/authorized_keys</code> on my server.</p>
<p>In my github action I’ve writing the private key as a file.</p>
<pre><code>-name: Create SSH key
run: echo "$GITHUB_PRIVATE_KEY" &gt; ../github_do.key
shell: bash
env:
GITHUB\_PRIVATE\_KEY: ${{secrets.GITHUB_PRIVATE_KEY}} 
</code></pre>
<p> Then using this file </p>
<pre><code>ssh -i path-to-file  ***@***"mkdir -p /home/ ***/*** /releases/2020-02-29-v***.244.0"
</code></pre>
<p>It fails on  <strong>Host key verification failed</strong>.</p>

**Comment by airtower-luna** [comment]
<p>That message means that the connection failed because your server could not be authenticated. You should write the host key (from /etc/ssh/ssh_host_KEYTYPE_key.pub) to the ~/.ssh/known_hosts file on the runner together with the host name, e.g.:</p>
<pre><code>server.example.net ssh-ed25519 AAAAC...
</code></pre>

**Comment by felixmosh** [comment]
<p>Thank you!</p>
<p>I’ve used this in order to add my server to known_hosts fille</p>
<pre><code>ssh-keyscan -H ${{secrets.DEPLOY_SERVER}} &gt; ~/.ssh/known_hosts
</code></pre>
<p>And it solved the previous error, but, now I get this:</p>
<pre><code>@ ***-err*** @***: Permission denied (publickey,gssapi-keyex,gssapi-with-mic).
</code></pre>
<p>Any pointers? </p>

**Comment by airtower-luna** [comment]
<p>That error probably means that your server doesn’t accept the configured private key for the target user. Check that the matching public key is included in the authorized_keys file of that user.</p>
<p>Also, if you use the result of ssh-keyscan without verification you’re effectively disabling host key checking. Consider if that really is an acceptable risk according to your security model.</p>

**Comment by felixmosh** [comment]
<p>Ok, i will try to with your approach.</p>

**Comment by felixmosh** [comment]
<p>Ok, so I’ve changed my approach,</p>
<p>I’ve entered to my server, <code>ssh-keyscan server.ip</code>, copied the first part to a secret, DO_GITHUB_PUBLIC_KEY.<br>
I’ve added this part to prepare known_hosts.</p>
<pre><code>echo "$DO_GITHUB_PUBLIC_KEY" &gt; ~/.ssh/known_hosts
</code></pre>
<p>And I still get the same thing :\</p>
<pre><code>@ ***-err*** @***: Permission denied (publickey,gssapi-keyex,gssapi-with-mic).
</code></pre>

**Comment by felixmosh** [comment, marked as answer]
<p>I solved it!!!<br>
Apparently keys were protected with passphrase :exploding_head:.</p>
<p>Once I’ve generated a new ones without it, it worked.</p>
<p>So that is the whole process:</p>
<ol>
<li>
<p>Genereate new keys</p>
<p>ssh-keygen -t rsa -b 4096 -C “user@host” -q -N “”</p>
</li>
<li>
<p>Update your host’s <code>authorized_keys</code></p>
<p>ssh-copy-id -i ~/.ssh/id_rsa.pub user@host</p>
</li>
<li>
<p>Enter the server &amp; run </p>
<p>ssh-keyscan host</p>
</li>
<li>
<p>Copy the output to github secret (let call it SSH_KNOWN_HOSTS)</p>
</li>
<li>
<p>Copy the <strong>private</strong> key to a github secret (lets call it SSH_PRIVATE_KEY)</p>
</li>
</ol>
<p>In your workflow.yml file</p>
<pre><code>jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - name: Create SSH key
        run: |
          mkdir -p ~/.ssh/
          echo "$SSH_PRIVATE_KEY" &gt; ../private.key
          sudo chmod 600 ../private.key
          echo "$SSH_KNOWN_HOSTS" &gt; ~/.ssh/known_hosts
        shell: bash
        env:
          SSH_PRIVATE_KEY: ${{secrets.SSH_PRIVATE_KEY}}
          SSH_KNOWN_HOSTS: ${{secrets.SSH_KNOWN_HOSTS}}
</code></pre>
<p>Then you can use ssh with </p>
<pre><code>ssh -i path-to/private.key user@host
</code></pre>

**Reply by ngohuytrieu** [reply]
I run into this issue today
Here is my action
```
- name: Create SSH key
        run: |
          mkdir -p ~/.ssh/
          echo "$SSH_PRIVATE_KEY" > ~/.ssh/private.key
          sudo chmod 600 ~/.ssh/private.key
          echo "$SSH_KNOWN_HOSTS" > ~/.ssh/known_hosts
          sudo chmod 600 ~/.ssh/known_hosts
        shell: bash
        env:
          SSH_PRIVATE_KEY: ${{secrets.SSH_KEY}}
          SSH_KNOWN_HOSTS: ${{secrets.SSH_HOSTNAME}}
      - name: SSH into DO and run
        run: >
          ssh -i ~/.ssh/private.key -tt username@hostname
          "
            echo 'Welcome to ec2'
          "
```

Any idea @felixmosh ?

**Reply by felixmosh** [reply]
Did you read what I've written above?
Make sure that you don't have passphrase on the keys

**Comment by virtuallabdev** [comment]
<p>Could you please provide an example with rsync? thanks in advance</p>

**Comment by airtower-luna** [comment]
<p>What are you trying to do with rsync? There shouldn’t be any difference, if SSH works you can use rsync over SSH without any additional steps.</p>

**Comment by felixmosh** [comment]
<p><code>rsync</code> works with <code>ssh</code>, like <a class="mention" href="/airtower-luna">@airtower-luna</a> said, all you need is a working ssh…</p>
